### ISSUE TITLE
Hot Reload not working (on Web and Linux)

### ISSUE THREAD

**Issue by dxps** [DioxusLabs/dioxus#1029]
**Problem**

_Hot reload_ does not work using Web. Not even after a refresh (F5).
However, the WebSocket established connection seems to push the updates to the UI.

**Steps To Reproduce**

- Just running with `dioxus serve --hot-reload` in [this project](https://github.com/dxps/fullstack-rust-axum-dioxus-rwa/tree/main/frontend).

**Expected behavior**

Expected that _hot reload_ should just work (as before, in 0.2.x) using Web.

**Screenshots**

**Environment:**

 - Dioxus version:  `0.3.2`
 - Dioxus CLI version: `0.3.1`
 - Rust version: `1.69.0`
 - OS info: Pop!_OS 22.04 LTS (Linux distro)
 - App platform: `web`

---

** Additional Details **

According to [hot reload guide](https://github.com/DioxusLabs/dioxus/blob/6acec6f020c6d45c86dd6597a6f4507e731189a1/docs/guide/src/en/getting_started/hot_reload.md#L4), I also tried to add `dioxus-hot-reload` (ver. 0.1.1) and included the `hot_reload_init!();` before `dioxus_web::launch(App);`. 

But in this case, the compilation of `interprocess` - a transitive dependency - fails.
Also, raised [a ticket](https://github.com/kotauskas/interprocess/issues/37) to that crate.

```
…/frontend on  main [!] ❯ dioxus serve --hot-reload
[INFO] 🚅 Running build command...
- 💼 Waiting to start build the project...  

... (content omitted) ...

   Compiling notify v5.1.0
| Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               \ Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               - Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                                  Compiling ignore v0.4.20
error: could not compile `interprocess` due to 15 previous errors
warning: build failed, waiting for other jobs to finish...
/ Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               - Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               - Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               | Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               / Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               / Compiling to_method 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)                                                                               Error: 🚫 Serving project failed: error: Your target operating system is not supported by interprocess – check if yours is in the list of supported systems, and if not, please open an issue on the GitHub repository if you think that it should be included
  --> /home/dxps/.cargo/registry/src/github.com-1ecc6299db9ec823/interprocess-1.2.1/src/lib.rs:82:1
   |
82 | compile_error!("Your target operating system is not supported by interprocess – check if yours is in the list of supported systems, and if not, please open an issue on the GitHub repository if you think that it should be included...
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Thanks!

**Comment by DogeDark** [comment]
On web, you do not need dioxus-hot-reload or the ``--hot-reload`` flag. I cloned your project and using ``dioxus serve`` on its own works and has hot reloading, albeit a little slow (I imagine because its a large project).

**Comment by dxps** [comment]
Oh, dear! :man_facepalming: 
Iirc, I had to add the `--hot-reload` flag after upgrading to Dioxus 0.3, and with that flag it worked.
Then I believe it stopped working meanwhile. Anyway, good to know, and thanks @DogeDark for the feedback! :vulcan_salute: 

Yes, it works with just `dioxus serve`.
And we can agree that's not actually hot reload, but a full page refresh. Still very helpful though.

What's also interesting is that the UI tries to establish a WS channel for the hot reload, which is not available without that flag.

![image](https://github.com/DioxusLabs/dioxus/assets/14140226/0a4bc0cd-a060-4374-8dd2-41acc8e59367)

**Comment by dxps** [comment]
I'll close the ticket. And maybe we'll clarify later what's all about the hot reload, at least for the Web.

**Comment by DogeDark** [comment]
Yeah, I'm not most familiar with how Dioxus handles it all and you're correct it does seem to be a page reload - maybe you do need the hot reload flag?

**Comment by Demonthos** [comment]
The `--hot-reload` flag enables rsx-hot-reloading. If you change a static part of the rsx in your project, the Dioxus CLI can reload it without recompiling the project. It is difficult to tell what the issue here is without more details about what code you are trying to change and what the messages Dioxus sends in the websocket are. If you could provide those details, that would be very helpful.

There have been some bugfixes for hot reloading specific to Linux since the last release (#842), and some issues with workspaces. If you upgrade to the git version of Dioxus and the CLI, the issue may already be fixed.

**Comment by dxps** [comment]
@Demonthos Sure, I'm glad to help.

The context is this front-end (which for now is pretty minimal). The behavior is as follows:

| Using `dioxus serve` | Using `dioxus serve --hot-reload` |
| --- | --- |
| Updating a page component's code triggers a full page reload, a "reload" message is received on the ws channel, and the updates are reflected in the UI.<br/>    | There is no full page reload triggered as a side effect, even if the ws channel is established and the update is being received on the browser through the hot_reload channel.<br/>But nothing is received on the ws channel, so I assume that's the root cause of the issue. |
| ![image](https://github.com/DioxusLabs/dioxus/assets/14140226/2338a0d1-b691-4e56-aba2-e8568fd4ccfe)  | ![image](https://github.com/DioxusLabs/dioxus/assets/14140226/35e9d6b3-98d1-4b38-8bce-695c2497cb3b) |

**Comment by Demonthos** [comment]
> @Demonthos Sure, I'm glad to help.

Thanks for the screenshots, they were helpful in tracking down the real issue here.

The CLI is handling workspace paths incorrectly with the file macro. This causes the id of the template (which is based on the path) to be different in the real template and the updated template. Since the template id that is sent by the CLI is incorrect and found nowhere in the page, it causes no change.

This issue was fixed in #917, but the CLI relies on the published version of dioxus-rsx, so it has not received the bugfix yet

**Comment by dxps** [comment]
Alrighty. And thanks again for the feedback @Demonthos! :pray: 
Good to know about it and that it will be solved. :wink:
